hazelcast/src/main/java/com/hazelcast/nio/tcp/nonblocking/AbstractHandler.java

```diff
@@ -98,11 +102,33 @@ final void unregisterOp(int operation) throws IOException {
         selectionKey.interestOps(selectionKey.interestOps() & ~operation);
     }
 
+    @SuppressWarnings({"checkstyle:npathcomplexity"})
     @Override
     public void onFailure(Throwable e) {
         if (e instanceof OutOfMemoryError) {
             ioService.onOutOfMemory((OutOfMemoryError) e);
         }
+
+        // if our selection key was cancelled due to replacement of the selector
+        // (see NonBlockingIOThread#selectWorkaround), then reschedule this handler for execution
+        if (e instanceof CancelledKeyException && selectionKey != null
+                && selectionKey.selector() != ioThread.getSelector()) {
+            // the selector was replaced, so obtain a new key for the new selector
```

Comment:
@pveentjer If a migration is in progress, perhaps we risk overwriting the new `selectionKey` created in the new owner thread executing `completeMigration`. Should this condition be amended as follows to avoid interfering with a migration in progress? In this scenario, `this.ioThread` has already been substituted by `startMigration` with the new owner thread, while `onFailure` would still be executing a task on the previous owner.
```java
if (e instanceof CancelledKeyException && selectionKey != null
                && ioThread == Thread.currentThread()
                && selectionKey.selector() != ioThread.getSelector()) {
```